app/controllers/application_controller.rb

```diff
@@ -188,9 +188,33 @@ def service_provider_mfa_setup_url
     service_provider_mfa_policy.user_needs_sp_auth_method_setup? ? two_factor_options_url : nil
   end
 
+  def fix_broken_personal_key_url
+    if current_user.broken_personal_key?
+      pii_unlocked = user_session[:decrypted_pii].present?
+
+      if pii_unlocked
+        cacher = Pii::Cacher.new(current_user, user_session)
+        user_session[:personal_key] = current_user.active_profile.encrypt_recovery_pii(cacher.fetch)
+
+        flash[:info] = t('account.personal_key.needs_new')
+        analytics.track_event(Analytics::BROKEN_PERSONAL_KEY_REGENERATED)
+
+        manage_personal_key_url
+      else
+        flash[:info] = t('account.personal_key.needs_new_password_prompt')
+        user_session[:needs_new_personal_key] = true
+
+        capture_password_url
+      end
+    end
+  end
```

Comment:
I'm not in love with a plain "_url" method having a ton of side effects, especially in such a hot path... but I wasn't sure if we had better places to insert something into the "flow" after signing in but before getting redirected elsewhere